File: src/leetCodeProblems/TwoPointers/ContainerWithMostWater11.java
package leetCodeProblems.TwoPointers;

public class MaxAreaWaterContainer11 {
	
	public int maxArea(int[] A) {
        int leftPointer = 0;
        int rightPointer = A.length - 1;
        int areaMax = 0;
        
        while(leftPointer < rightPointer) {
        
            areaMax = Math.max(areaMax, (rightPointer - leftPointer) * Math.min(A[leftPointer], A[rightPointer]));
            
            if (A[leftPointer] < A[rightPointer]) {
                leftPointer += 1;
            }
            else {
                rightPointer -= 1;
            }
            
        }
        
        return areaMax;
    }

}

File: src/leetCodeProblems/TwoPointers/ContainerWithMostWater11.go
package main

/*
- Leetcode - https://leetcode.com/problems/container-with-most-water/description/
- Space - O(1)
- Time - O(n)
*/

func maxArea(height []int) int {
	out := 0
	left := 0
	right := len(height) - 1

	for left < right {
		area := min(height[left], height[right]) * (right - left)
		out = max(out, area)

		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}
	return out
}
